File: packages/neft-active-link/native/android/ActiveLinkExtension.java
package io.neft.extensions.activelink_extension;

import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;

import java.util.Locale;

import io.neft.App;
import io.neft.utils.Consumer;

public final class ActiveLinkExtension {
    private final static App APP = App.getInstance();

    public static void register() {
        APP.getClient().addCustomFunction("NeftActiveLink/mailto", new Consumer<Object[]>() {
            @Override
            public void accept(Object[] var) {
                mailto((String) var[0]);
            }
        });

        APP.getClient().addCustomFunction("NeftActiveLink/tel", new Consumer<Object[]>() {
            @Override
            public void accept(Object[] var) {
                tel((String) var[0]);
            }
        });

        APP.getClient().addCustomFunction("NeftActiveLink/geo", new Consumer<Object[]>() {
            @Override
            public void accept(Object[] var) {
                geo(((Number) var[0]).floatValue(), ((Number) var[1]).floatValue(), (String) var[2]);
            }
        });
    }

    public static void mailto(String address) {
        Intent intent = new Intent(Intent.ACTION_SENDTO);
        intent.setData(Uri.parse("mailto:" + address));
        try {
            APP.getActivity().startActivity(intent);
        } catch (ActivityNotFoundException error) {
            Log.e("NEFT", "Cannot mailto in ActiveLink extension", error);
        }
    }

    public static void tel(String number) {
        Intent intent = new Intent(Intent.ACTION_DIAL);
        intent.setData(Uri.parse("tel:" + number));
        try {
            APP.getActivity().startActivity(intent);
        } catch (ActivityNotFoundException error) {
            Log.e("NEFT", "Cannot tel in ActiveLink extension", error);
        }
    }

    public static void geo(Float latitude, Float longitude, String address) {
        String uri = String.format(Locale.ENGLISH, "geo:%f,%f?q=%s", latitude, longitude, address);
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
        try {
            APP.getActivity().startActivity(intent);
        } catch (ActivityNotFoundException error) {
            Log.e("NEFT", "Cannot geo in ActiveLink extension", error);
        }
    }
}

File: packages/lib/neft-active-link/native/android/ActiveLinkExtension.kt
package io.neft.extensions.activelink_extension

import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.util.Log

import io.neft.App
import io.neft.client.NativeBinding

object ActiveLinkExtension {
    private val APP = App.getInstance()
    private val BINDING = NativeBinding("ActiveLink")

    @JvmStatic
    fun register() {
        BINDING
                .onCall("web") { args: Array<Any> ->
                    web(args[0] as String)
                }
                .onCall("mailto") { args: Array<Any> ->
                    mailto(args[0] as String)
                }
                .onCall("tel") { args: Array<Any> ->
                    tel(args[0] as String)
                }
                .onCall("geo") { args: Array<Any> ->
                    geo(
                            latitude = (args[0] as Number).toFloat(),
                            longitude = (args[1] as Number).toFloat(),
                            address = args[2] as String
                    )
                }
    }

    fun web(url: String) {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        try {
            APP.activity.startActivity(intent)
        } catch (error: ActivityNotFoundException) {
            Log.e(App.TAG, "Cannot web in ActiveLink extension", error)
        }
    }

    fun mailto(address: String) {
        val intent = Intent(Intent.ACTION_SENDTO)
        intent.data = Uri.parse("mailto:$address")
        try {
            APP.activity.startActivity(intent)
        } catch (error: ActivityNotFoundException) {
            Log.e(App.TAG, "Cannot mailto in ActiveLink extension", error)
        }
    }

    fun tel(number: String) {
        val intent = Intent(Intent.ACTION_DIAL)
        intent.data = Uri.parse("tel:$number")
        try {
            APP.activity.startActivity(intent)
        } catch (error: ActivityNotFoundException) {
            Log.e(App.TAG, "Cannot tel in ActiveLink extension", error)
        }
    }

    fun geo(latitude: Float?, longitude: Float?, address: String) {
        val uri = "geo:$latitude,$longitude?q=$address"
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(uri))
        try {
            APP.activity.startActivity(intent)
        } catch (error: ActivityNotFoundException) {
            Log.e(App.TAG, "Cannot geo in ActiveLink extension", error)
        }
    }
}
